### PR TITLE
net: tcp: When closing the connection send FIN immediately

### LINKED ISSUE #19678: slowness networking stack - MXRT1050-EVK
I'm trying the "samples/net/sockets/dumb_http_server", i just developed a little webserver based on this example.
The request always takes about 350ms, you can see it from the browser console (F12 -> network).
I think the problem is not related to the board or network because my bare-metal version is much faster (the HTTP respose is like 1ms).

I think the problem is in the "accept" call, but I don't know if is the context switch delay from the "idle" thread or other.

Can anyone reproduce the problem? I'm using the MIMXRT1050-EVK.

### PR DESCRIPTION
**Pull request by jukkar**
The earlier code was queuing the FIN that is sent when connection
is closed. This caused long delay (200 ms) before the peer at the
other end noticed that the connection was actually closed.

Fixes #19678

Signed-off-by: Jukka Rissanen <jukka.rissanen@linux.intel.com>

### CODE REVIEW

**Review by tbursztyka — APPROVED** [review]

**Comment by jukkar** [comment]
I will debug this a bit more, apache bench reports peer resets.

**Comment by pfalcon** [comment]
> This caused long delay (200 ms) before the peer at the
other end noticed that the connection was actually closed.

If you debugged and found that cause, kudos. In my list, that 200+ms delay appeared after last @tbursztyka's last refactor (7f2cb027204b64fc63320db0ed444b3150c87606 and friends) (note: not pointing fingers, but providing information/context). I should have reported it, though can't find a reference (e.g., it's not in https://github.com/zephyrproject-rtos/zephyr/issues/13943, which is another issue happened soon after refactor).


But with the in mind, this:

> net: tcp: When closing the connection send FIN immediately. The earlier code was queuing the FIN that is sent when connection is closed.

 `-` is not correct. FIN is just as TCP segment as any other (e.g. a data segment). Actually, all TCP segments are the same. They should be queued. They should be retransmitted if response doesn't arrive at time. They should be sent in proper order. This starts with SYN segment - it should be sent via standard TCP means, in particular, retransmitted if no reply. We handle it an adhoc way, and due to this don't retransmit, which is a known issue (https://github.com/zephyrproject-rtos/zephyr/issues/9944).

The same applies to FIN: it should be put in the TCP segment queue. Which may contain TCP data segments not fully sent yet. FIN should be sent only after they are sent out. But they may be retransmitted yet, so there should be no hurry to send FIN. Finally, FIN itself might be retransmitted a few times.


Anyway, writing that based on the description here in the PR, let me look at the code.

**Review by pfalcon — CHANGES_REQUESTED** [review]
So yeah, based on the previous comment (https://github.com/zephyrproject-rtos/zephyr/pull/19747#issuecomment-540633363), this is not a correct change and I have to give -1.

FIN really should be queued. But it shouldn't linger there without a reason. Let's figure out what causes that 200ms with it being sent out (or whatever the actual underlying issue may be).

**Comment by jukkar** [comment]
> If you debugged and found that cause, kudos. In my list, that 200+ms delay appeared after last @tbursztyka's last refactor ([7f2cb02](https://github.com/zephyrproject-rtos/zephyr/commit/7f2cb027204b64fc63320db0ed444b3150c87606) and friends) (note: not pointing fingers, but providing information/context). I should have reported it, though can't find a reference (e.g., it's not in #13943, which is another issue happened soon after refactor).

Difficult to know the culprit commit. Last commit that changed things around the active close was your commit 338dc8a9520ea7b3c60503e35daba085073f4a1d long time ago.

File: subsys/net/ip/tcp.c [hunk 1]
```diff
@@ -972,6 +972,11 @@ int net_tcp_send_pkt(struct net_pkt *pkt)
 	return net_send_data(pkt);
 }
 
+static void flush_queue(struct net_context *context)
+{
+	(void)net_tcp_send_data(context, NULL, NULL);
```
Comment (pfalcon) [inline subsys/net/ip/tcp.c]: Comment to this function says:

~~~
        /* For now, just send all queued data synchronously.  Need to
         * add window handling and retry/ACK logic.
         */
~~~

So, it still would be a workaround, not a solution, and could lead to silent data loss.

**Comment by pfalcon** [comment]
Tested this with standard test method I used for years (qemu_x86+SLIP, dump_http_server, ab -n1000):

~~~
Percentage of the requests served within a certain time (ms)
  50%     22
  66%     25
  75%     29
  80%     33
  90%   1026
  95%   1035
  98%   1047
  99%   1051
 100%   1069 (longest request)
~~~

That IMHO only deteriorates the situation, as instead of stable ~200ms we get high delay dispersion of unknown reason. So, I'll keep my -1 and extend my call to understand where 200ms delays comes from (while it shouldn't be there), instead of throwing random workaround.

(Mostly a crosspost from https://github.com/zephyrproject-rtos/zephyr/issues/19678#issuecomment-541021757)

**Comment by pfalcon** [comment]
As was found in https://github.com/zephyrproject-rtos/zephyr/issues/19678#issuecomment-541243305 , this effectively tries to undo the effect of https://github.com/zephyrproject-rtos/zephyr/commit/45df09c010d29684a0f2fe7e2b4751c9cea25b48 . And arguably, it does it in a roundabout way, which still raises questions how correct it is (i.e. what are corner cases and unobvious implications of doing it this way).

I'd suggest to seriously consider reinstating the previous code as another option (while of course addressing @dleach02's concern as described in the above commit message). Or alternatively, explain in more detail why the solution here is better. Note that such an explanation would need to account for the actual test data: https://github.com/zephyrproject-rtos/zephyr/issues/19678#issuecomment-541021757 , which shows that with this patch, there's noticeable and high-dispersion delay distribution of unknown nature for connection handling times, whereas simply reverting 45df09c010d29684a0f2fe7e2b4751c9cea25b48 would lead to a nice flat delay distribution with flat dispersion.

**Comment by jukkar** [comment]
@pfalcon thanks for thorough investigation about the issue. After looking the code, I think we should not revert commit 45df09c as the original code was doing a bit suspicious things. It for example did not check the 6lo tech check that is done in `net_tcp_send_data()` when calling `net_tcp_send_pkt()`.

Another suspicious stuff that current code does is in https://github.com/zephyrproject-rtos/zephyr/blob/bb9c453777021cb0cd8c0ea4b50adfe7dc2a1bf7/subsys/net/ip/tcp.c#L1017-L1024
The function does not return if he packet was not sent, but continues to call `net_pkt_set_queued(pkt, true);` which does not look right.

**Comment by jukkar** [comment]
I think what this PR proposes is more or less correct thing to do:
* If the there are no other TCP segments queued to be sent, it then sends the final FIN immediately
* If there are some data segments to be sent, the FIN is queued and sent after those ones
* After receiving a valid ack, we flush the queue in order to send any pending FIN

**Review by pfalcon — APPROVED** [review]
@pfalcon wrote:

>   50%     22
>  90%   1026
 
> That IMHO only deteriorates the situation, as instead of stable ~200ms we get high delay dispersion of unknown reason.

@jukkar wrote in https://github.com/zephyrproject-rtos/zephyr/issues/19678#issuecomment-542186844:

> Sample: dump_http_server
Git: master + fix #19747
Results: The request take about 1000ms. Because we are no longer waiting 200ms, ab gets results quicker and can send new request earlier. The dumb_http_server is a single threaded server which will wait for incoming connections, serve them and then wait another connection. At the moment we do not have any queing for received SYN packets so if the application has not called accept(), the incoming SYN is dropped. This is happening here and I verified it from debug logs. The sample app is a bit too slow to accept another connection attempt and this forces the Linux to re-send a new SYN packet after around 1sec timeout. This is not happening always, sometimes dumb_http_server manages to accept the connection in time, thus the ab reported time is <1sec.

To verify this, I took this patch and applied revert for https://github.com/zephyrproject-rtos/zephyr/commit/1a6f4a6368960ddcbeb776f45f084163c7001f36 (required a trivial fixup).

After that, testing mps2_an385+overlay-smsc911x.conf (again, to avoid QEMU+SLIP which are known to have issues) leads to:

~~~
$ time ab -n1000 http://192.0.2.1:8080/
...
Percentage of the requests served within a certain time (ms)
  50%     12
  66%     13
  75%     14
  80%     14
  90%     15
  95%     16
  98%     17
  99%     18
 100%     23 (longest request)
~~~

@jukkar 
> I think what this PR proposes is more or less correct thing to do:

With the all explanations provided and my testing above which adheres to the explanations, I may only agree with the above, this is definitely an improvement over what we had before, and itself doesn't seem to introduce more issues (but exposes issues we got independently some time ago).

And in a long while, this a case where there was actual thorough testing (with multiple targets, etc.) **and** results of that testing were documented and discussed/explained, I very much appreciate that.

**Pull request merged**